File: src/jarvis/remote/origin.py
"""Canonical server-origin handling for authority-bound mobile identities."""

from __future__ import annotations

import ipaddress
import re
from contextlib import suppress
from urllib.parse import urlsplit

_DNS_LABEL = re.compile(r"^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$")


def normalize_server_origin(value: str, *, allow_insecure_loopback: bool = False) -> str:
    """Return one unambiguous origin or reject unsafe/non-origin URLs."""
    if value != value.strip() or len(value) > 255:
        raise ValueError("server origin must be trimmed and at most 255 characters")
    parsed = urlsplit(value)
    scheme = parsed.scheme.casefold()
    if parsed.username is not None or parsed.password is not None:
        raise ValueError("server origin cannot contain user information")
    if parsed.path not in {"", "/"} or parsed.query or parsed.fragment:
        raise ValueError("server origin cannot contain a path, query, or fragment")
    hostname = parsed.hostname
    if hostname is None or not hostname.isascii() or hostname.endswith("."):
        raise ValueError("server origin requires an ASCII host without a trailing dot")
    hostname = hostname.casefold()
    try:
        ipaddress.ip_address(hostname)
    except ValueError:
        labels = hostname.split(".")
        if any(_DNS_LABEL.fullmatch(label) is None for label in labels):
            raise ValueError("server origin contains an invalid host") from None
    try:
        port = parsed.port
    except ValueError as exc:
        raise ValueError("server origin contains an invalid port") from exc
    if port == 0:
        raise ValueError("server origin contains an invalid port")
    loopback = hostname == "localhost"
    with suppress(ValueError):
        loopback = loopback or ipaddress.ip_address(hostname).is_loopback
    if scheme != "https" and not (scheme == "http" and allow_insecure_loopback and loopback):
        raise ValueError("server origin must use HTTPS")
    default_port = 443 if scheme == "https" else 80
    rendered_host = f"[{hostname}]" if ":" in hostname else hostname
    authority = rendered_host if port in {None, default_port} else f"{rendered_host}:{port}"
    return f"{scheme}://{authority}"


def authority_from_origin(origin: str) -> str:
    """Extract canonical authority from an already-normalized server origin."""
    parsed = urlsplit(normalize_server_origin(origin, allow_insecure_loopback=True))
    return parsed.netloc

File: mobile/src/core/auth/origin.ts
const dnsLabel = /^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/;

export function normalizeServerOrigin(
  value: string,
  options: Readonly<{ allowInsecureLoopback?: boolean }> = {},
): string {
  if (value !== value.trim() || value.length === 0 || value.length > 255) {
    throw new Error("server origin must be trimmed and at most 255 characters");
  }
  if (!/^[\x00-\x7f]+$/.test(value)) {
    throw new Error("server origin must contain only ASCII characters");
  }
  let url: URL;
  try {
    url = new URL(value);
  } catch {
    throw new Error("server origin is invalid");
  }
  if (url.username || url.password || url.pathname !== "/" || url.search || url.hash) {
    throw new Error("server origin cannot contain credentials, path, query, or fragment");
  }
  const hostname = url.hostname.toLowerCase();
  if (!hostname || hostname.endsWith(".")) {
    throw new Error("server origin contains an invalid host");
  }
  const ipv6 = hostname.startsWith("[") && hostname.endsWith("]");
  const ipv4 = /^\d{1,3}(?:\.\d{1,3}){3}$/.test(hostname);
  if (
    (!ipv6 && !ipv4 && hostname.split(".").some((label) => !dnsLabel.test(label))) ||
    (ipv6 && !/^\[[0-9a-f:]+\]$/.test(hostname))
  ) {
    throw new Error("server origin contains an invalid host");
  }
  if (ipv4 && hostname.split(".").some((part) => Number(part) > 255)) {
    throw new Error("server origin contains an invalid host");
  }
  if (url.port === "0") {
    throw new Error("server origin contains an invalid port");
  }
  const loopback = hostname === "localhost" || hostname.startsWith("127.") || hostname === "[::1]";
  if (
    url.protocol !== "https:" &&
    !(url.protocol === "http:" && options.allowInsecureLoopback === true && loopback)
  ) {
    throw new Error("server origin must use HTTPS");
  }
  return `${url.protocol}//${url.host.toLowerCase()}`;
}
